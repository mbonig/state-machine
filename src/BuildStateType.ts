export function buildStateType(stateDef: any, indent = 2) {
  let stateType = 'Partial<{\n';
  const indentTabs = '  '.repeat(indent);
  const closeIndents = '  '.repeat(indent - 1);
  if (stateDef === null) {
    // we're going to return an any here because while it's null now, maybe it'll be an any later.
    return 'any';
  }
  if (typeof stateDef === 'string') {
    return 'string';
  }
  if (Number.isInteger(stateDef)) {
    return 'number';
  }
  for (const prop of Object.keys(stateDef)) {
    const propValue = stateDef[prop];
    const safeProp = prop.replace(/'/g, '\\\'');
    if (typeof propValue === 'string') {
      stateType += `${indentTabs}'${safeProp}': string;\n`;
    } else if (Array.isArray(propValue)) {
      const arrayPropTypes = propValue.map(pv => buildStateType(pv, indent + 1));
      stateType += `${indentTabs}'${safeProp}': Partial<[${arrayPropTypes.join(', ')}]>;\n`;
    } else {
      stateType += `${indentTabs}'${safeProp}': ${buildStateType(propValue, indent + 1)};\n`;
    }
  }
  stateType += `${closeIndents}}>`;
  return stateType;
}
