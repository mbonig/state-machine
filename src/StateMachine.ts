import { Duration } from 'aws-cdk-lib';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as aws_stepfunctions from 'aws-cdk-lib/aws-stepfunctions';
import { CfnStateMachine, LogOptions, Pass, StateMachineType } from 'aws-cdk-lib/aws-stepfunctions';
import { Construct } from 'constructs';
// eslint-disable-next-line @typescript-eslint/no-require-imports
const merge = require('lodash.merge');

export interface StateMachineProps {

  /**
   * An object that matches the schema/shape of the ASL .States map with overridden values.
   *
   * @example {'My First State': { Parameters: { FunctionName: 'aLambdaFunctionArn' } } }
   */
  readonly overrides?: any;

  /**
   * An object that can be serialized into an ASL.
   */
  readonly definition: any;

  /**
   * A name for the state machine
   *
   * @default A name is automatically generated
   */
  readonly stateMachineName?: string;

  /**
   * The execution role for the state machine service
   *
   * @default A role is automatically created
   */
  readonly role?: iam.IRole;

  /**
   * Maximum run time for this state machine
   *
   * @default No timeout
   */
  readonly timeout?: Duration;

  /**
   * Type of the state machine
   *
   * @default StateMachineType.STANDARD
   */
  readonly stateMachineType?: StateMachineType;

  /**
   * Defines what execution history events are logged and where they are logged.
   *
   * @default No logging
   */
  readonly logs?: LogOptions;

  /**
   * Specifies whether Amazon X-Ray tracing is enabled for this state machine.
   *
   * @default false
   */
  readonly tracingEnabled?: boolean;

}

export class StateMachine extends aws_stepfunctions.StateMachine {

  constructor(scope: Construct, id: string, props: StateMachineProps) {
    super(scope, id, {
      ...props,
      definition: new Pass(scope, 'THISWILLBEDELETEDRIGHTAWAY'),
    });
    scope.node.tryRemoveChild('THISWILLBEDELETEDRIGHTAWAY');
    const mergedDefinition = merge(props.definition, { States: props.overrides });
    (this.node.defaultChild as CfnStateMachine).definitionString = JSON.stringify(mergedDefinition);
  }
}

