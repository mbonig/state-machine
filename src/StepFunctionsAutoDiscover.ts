import * as fs from 'fs';
import * as path from 'path';
import { basename, dirname, extname, join } from 'path';
import { pascal } from 'case';
import * as yaml from 'js-yaml';
import { Component, Project, SourceCode } from 'projen';
import { AwsCdkTypeScriptApp } from 'projen/lib/awscdk';
import { AwsCdkDeps } from 'projen/lib/awscdk/awscdk-deps';
import { AutoDiscoverBase } from 'projen/lib/cdk';
import { buildStateType } from './BuildStateType';

/**
 * The original extension used to search for ASL files.
 */
export const JSON_STEPFUNCTION_EXT = '.workflow.json';

/**
 * The AWS-recommended extension for ASL files.
 */
export const AWS_RECOMMENDED_JSON_EXT = '.json.asl';

/**
 * The AWS-recommended extension for YAML ASL files.
 */
export const AWS_RECOMMENDED_YAML_EXT = '.yaml.asl';

export interface StepFunctionsAutoDiscoverOptions {
  /**
   * An optional extension to use for discovering state machine files.
   *
   * @default '.workflow.json' (JSON_STEPFUNCTION_EXT)
   */
  readonly extension?: string;
}

/**
 * A projen component for discovering AWS Step Function state machine workflow ASL files
 * and generating a strongly typed interface and construct to use it.
 *
 * Simply add a new instance and hand it your AwsCdkTypeScriptApp projen class:
 * ```
 * const project = new AwsCdkTypeScriptApp({ ... });
 * new StepFunctionsAutoDiscover(project);
 * ```
 *
 * And any *.workflow.json file will cause the generation of a new strongly-typed StateMachine-derived class you can use.
 * Note that these constructs are NOT jsii-compatible. If you need that,
 * please open an [issue](https://github.com/mbonig/state-machine/issues/new)
 */
export class StepFunctionsAutoDiscover extends AutoDiscoverBase {
  constructor(project: AwsCdkTypeScriptApp, _options?: StepFunctionsAutoDiscoverOptions) {
    if (_options?.extension) {
      if (!_options.extension.startsWith('.')) {
        throw new Error('extension must start with a .');
      }
    }
    let extension = _options?.extension || JSON_STEPFUNCTION_EXT;
    super(project, {
      extension: extension,
      projectdir: project.srcdir,
    });
    for (const entrypoint of this.entrypoints) {
      new StepFunctionsStateMachine(this.project, {
        workflowAsl: entrypoint,
        cdkDeps: project.cdkDeps,
        extension,
      });
    }
  }
}

export interface StepFunctionsStateMachineOptions {
  readonly extension: string;
  readonly constructFile?: string;
  readonly workflowAsl: string;
  readonly constructName?: string;
  readonly cdkDeps: AwsCdkDeps;
}

/**
 * Don't use this class directly.
 */
export class StepFunctionsStateMachine extends Component {
  constructor(project: Project, options: StepFunctionsStateMachineOptions) {
    super(project);

    const extension = options.extension ?? JSON_STEPFUNCTION_EXT;
    const workflowAsl = options.workflowAsl;

    if (
      !workflowAsl.endsWith(extension)
    ) {
      throw new Error(
        `${workflowAsl} must have a ${extension} extension`,
      );
    }
    const basePath = join(
      dirname(workflowAsl),
      basename(
        workflowAsl,
        extension,
      ),
    );
    const constructFile = options.constructFile ?? `${basePath}-statemachine.ts`;

    if (extname(constructFile) !== '.ts') {
      throw new Error(
        `Construct file name "${constructFile}" must have a .ts extension`,
      );
    }

    // type names
    const constructName =
      options.constructName ?? pascal(basename(basePath)) + 'StateMachine';

    const src = new SourceCode(project, constructFile);
    if (src.marker) {
      src.line(`// ${src.marker}`);
    }
    src.line('import fs from \'fs\';');
    src.line('import path from \'path\';');
    src.line('import { StateMachine, StateMachineProps } from \'@matthewbonig/state-machine\';');
    src.line('import { Construct } from \'constructs\';');

    let isYaml = /(yaml|yml)/.test(extension);
    if (isYaml) {
      src.line('import * as yaml from \'js-yaml\';');
    }

    src.open(`export interface ${constructName}Overrides {`);

    let workflowDefinition: any;
    if (isYaml) {
      workflowDefinition = yaml.load(fs.readFileSync(join(project.outdir, workflowAsl)).toString());
    } else {
      workflowDefinition = JSON.parse(fs.readFileSync(join(project.outdir, workflowAsl)).toString());

    }
    if (!workflowDefinition.States) {
      throw new Error(`The workflow file ${workflowAsl} doesn't appear to be a valid ASL file, it doesn't contain a 'States' field`);
    }
    for (const [stateName, stateDef] of Object.entries(workflowDefinition.States)) {
      const stateType = buildStateType(stateDef);
      src.line(`readonly '${stateName}': ${stateType};`);
    }
    src.close('}');

    src.open(`export interface ${constructName}Props extends Omit<StateMachineProps, 'overrides' | 'definition'> {`);
    src.line(`readonly overrides: Partial<${constructName}Overrides>;`);
    src.close('}');


    src.open(`export class ${constructName} extends StateMachine {`);
    src.open(`constructor(scope: Construct, id: string, props: ${constructName}Props) {`);
    src.open('super(scope, id, {');
    src.line('...props,');
    const relativeFile = path.basename(workflowAsl);
    if (isYaml) {
      src.line(`definition: yaml.load(fs.readFileSync(path.join(__dirname, '${relativeFile}')).toString()),`);
      src.line('aslYaml: true,');
    } else {
      src.line(`definition: JSON.parse(fs.readFileSync(path.join(__dirname, '${relativeFile}')).toString()),`);

    }
    src.close('});');
    src.close('}');
    src.close('}');

    this.project.logger.verbose(
      `${basePath}: construct "${constructName}" generated under "${constructFile}"`,
    );
  }
}
