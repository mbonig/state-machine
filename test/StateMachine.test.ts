import { App, Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { StateMachineType } from 'aws-cdk-lib/aws-stepfunctions';
import { StateMachine } from '../src';

test('simple snapshot', () => {
  const stack = new Stack(new App(), 'TestingStack', {});
  new StateMachine(stack, 'Test', {
    stateMachineName: 'A-nice-state-machine',
    overrides: { one: { Parameters: { thingOne: 'test passed' } } },
    stateMachineType: StateMachineType.STANDARD,
    definition: { States: { one: { Parameters: { thingOne: 'test failed' } } } },
  });
  const assert = Template.fromStack(stack);
  expect(assert.toJSON()).toMatchSnapshot();
});

test('handles arrays', () => {
  const stack = new Stack(new App(), 'TestingStack', {});
  let startingDefinition = {
    States: {
      Branches: [
        {
          StartAt: 'ResumeCluster',
          States: {
            'ResumeCluster': {
              Type: 'Task',
              Parameters: {
                ClusterIdentifier: 'MyData',
              },
              Resource: 'arn:aws:states:::aws-sdk:redshift:resumeCluster',
              Next: 'DescribeClusters',
            },
            'DescribeClusters': {
              Type: 'Task',
              Parameters: {
                ClusterIdentifier: '',
              },
              Resource: 'arn:aws:states:::aws-sdk:redshift:describeClusters',
              Next: 'Evaluate Cluster Status',
            },
            'Evaluate Cluster Status': {
              Type: 'Choice',
              Choices: [
                {
                  Variable: '$.Clusters[0].ClusterStatus',
                  StringEquals: 'available',
                  Next: 'Redshift Pass',
                },
              ],
              Default: 'Redshift Wait',
            },
            'Redshift Pass': {
              Type: 'Pass',
              End: true,
            },
            'Redshift Wait': {
              Type: 'Wait',
              Seconds: 5,
              Next: 'DescribeClusters',
            },
          },
        },
        {
          StartAt: 'StartInstances',
          States: {
            'StartInstances': {
              Type: 'Task',
              Parameters: {
                InstanceIds: [
                  'MyData',
                ],
              },
              Resource: 'arn:aws:states:::aws-sdk:ec2:startInstances',
              Next: 'DescribeInstanceStatus',
            },
            'DescribeInstanceStatus': {
              Type: 'Task',
              Next: 'Evaluate Instance Status',
              Parameters: {
                InstanceIds: [
                  'MyData',
                ],
              },
              Resource: 'arn:aws:states:::aws-sdk:ec2:describeInstanceStatus',
            },
            'Evaluate Instance Status': {
              Type: 'Choice',
              Choices: [
                {
                  And: [
                    {
                      Variable: '$.InstanceStatuses[0].InstanceState.Name',
                      StringEquals: 'running',
                    },
                    {
                      Variable: '$.InstanceStatuses[0].SystemStatus.Details[0].Status',
                      StringEquals: 'passed',
                    },
                    {
                      Variable: '$.InstanceStatuses[0].InstanceStatus.Details[0].Status',
                      StringEquals: 'passed',
                    },
                  ],
                  Next: 'EC2 Pass',
                },
              ],
              Default: 'EC2 Wait',
            },
            'EC2 Pass': {
              Type: 'Pass',
              End: true,
            },
            'EC2 Wait': {
              Type: 'Wait',
              Seconds: 5,
              Next: 'DescribeInstanceStatus',
            },
          },
        },
      ],
    },
  };
  new StateMachine(stack, 'Test', {
    stateMachineName: 'A-nice-state-machine',
    overrides: {
      Branches: [{
        StartAt: 'ResumeCluster',
        States: {
          ResumeCluster: {
            Parameters: {
              ClusterIdentifier: 'CLUSTER_NAME',
            },
          },
          DescribeClusters: {
            Parameters: {
              ClusterIdentifier: 'CLUSTER_NAME',
            },
          },
        },
      }, {
        StartAt: 'StartInstances',
        States: {
          StartInstances: {
            Parameters: {
              InstanceIds: ['INSTANCE_ID'],
            },
          },
          DescribeInstanceStatus: {
            Parameters: {
              InstanceIds: ['INSTANCE_ID'],
            },
          },
        },
      }],
    },
    stateMachineType: StateMachineType.STANDARD,
    definition: startingDefinition,
  });
  const assert = Template.fromStack(stack);
  assert.hasResourceProperties('AWS::StepFunctions::StateMachine', {
    DefinitionString: JSON.stringify({
      States: {
        Branches: [
          {
            StartAt: 'ResumeCluster',
            States: {
              'ResumeCluster': {
                Type: 'Task',
                Parameters: {
                  ClusterIdentifier: 'CLUSTER_NAME',
                },
                Resource: 'arn:aws:states:::aws-sdk:redshift:resumeCluster',
                Next: 'DescribeClusters',
              },
              'DescribeClusters': {
                Type: 'Task',
                Parameters: {
                  ClusterIdentifier: 'CLUSTER_NAME',
                },
                Resource: 'arn:aws:states:::aws-sdk:redshift:describeClusters',
                Next: 'Evaluate Cluster Status',
              },
              'Evaluate Cluster Status': {
                Type: 'Choice',
                Choices: [
                  {
                    Variable: '$.Clusters[0].ClusterStatus',
                    StringEquals: 'available',
                    Next: 'Redshift Pass',
                  },
                ],
                Default: 'Redshift Wait',
              },
              'Redshift Pass': {
                Type: 'Pass',
                End: true,
              },
              'Redshift Wait': {
                Type: 'Wait',
                Seconds: 5,
                Next: 'DescribeClusters',
              },
            },
          },
          {
            StartAt: 'StartInstances',
            States: {
              'StartInstances': {
                Type: 'Task',
                Parameters: {
                  InstanceIds: ['INSTANCE_ID'],
                },
                Resource: 'arn:aws:states:::aws-sdk:ec2:startInstances',
                Next: 'DescribeInstanceStatus',
              },
              'DescribeInstanceStatus': {
                Type: 'Task',
                Next: 'Evaluate Instance Status',
                Parameters: {
                  InstanceIds: ['INSTANCE_ID'],
                },
                Resource: 'arn:aws:states:::aws-sdk:ec2:describeInstanceStatus',
              },
              'Evaluate Instance Status': {
                Type: 'Choice',
                Choices: [
                  {
                    And: [
                      {
                        Variable: '$.InstanceStatuses[0].InstanceState.Name',
                        StringEquals: 'running',
                      },
                      {
                        Variable: '$.InstanceStatuses[0].SystemStatus.Details[0].Status',
                        StringEquals: 'passed',
                      },
                      {
                        Variable: '$.InstanceStatuses[0].InstanceStatus.Details[0].Status',
                        StringEquals: 'passed',
                      },
                    ],
                    Next: 'EC2 Pass',
                  },
                ],
                Default: 'EC2 Wait',
              },
              'EC2 Pass': {
                Type: 'Pass',
                End: true,
              },
              'EC2 Wait': {
                Type: 'Wait',
                Seconds: 5,
                Next: 'DescribeInstanceStatus',
              },
            },
          },
        ],
      },
    }),
  });
});
