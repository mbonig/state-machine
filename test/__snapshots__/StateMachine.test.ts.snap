// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`simple snapshot 1`] = `
Object {
  "Mappings": Object {
    "ServiceprincipalMap": Object {
      "af-south-1": Object {
        "states": "states.af-south-1.amazonaws.com",
      },
      "ap-east-1": Object {
        "states": "states.ap-east-1.amazonaws.com",
      },
      "ap-northeast-1": Object {
        "states": "states.ap-northeast-1.amazonaws.com",
      },
      "ap-northeast-2": Object {
        "states": "states.ap-northeast-2.amazonaws.com",
      },
      "ap-northeast-3": Object {
        "states": "states.ap-northeast-3.amazonaws.com",
      },
      "ap-south-1": Object {
        "states": "states.ap-south-1.amazonaws.com",
      },
      "ap-southeast-1": Object {
        "states": "states.ap-southeast-1.amazonaws.com",
      },
      "ap-southeast-2": Object {
        "states": "states.ap-southeast-2.amazonaws.com",
      },
      "ap-southeast-3": Object {
        "states": "states.ap-southeast-3.amazonaws.com",
      },
      "ca-central-1": Object {
        "states": "states.ca-central-1.amazonaws.com",
      },
      "cn-north-1": Object {
        "states": "states.cn-north-1.amazonaws.com",
      },
      "cn-northwest-1": Object {
        "states": "states.cn-northwest-1.amazonaws.com",
      },
      "eu-central-1": Object {
        "states": "states.eu-central-1.amazonaws.com",
      },
      "eu-north-1": Object {
        "states": "states.eu-north-1.amazonaws.com",
      },
      "eu-south-1": Object {
        "states": "states.eu-south-1.amazonaws.com",
      },
      "eu-south-2": Object {
        "states": "states.eu-south-2.amazonaws.com",
      },
      "eu-west-1": Object {
        "states": "states.eu-west-1.amazonaws.com",
      },
      "eu-west-2": Object {
        "states": "states.eu-west-2.amazonaws.com",
      },
      "eu-west-3": Object {
        "states": "states.eu-west-3.amazonaws.com",
      },
      "me-south-1": Object {
        "states": "states.me-south-1.amazonaws.com",
      },
      "sa-east-1": Object {
        "states": "states.sa-east-1.amazonaws.com",
      },
      "us-east-1": Object {
        "states": "states.us-east-1.amazonaws.com",
      },
      "us-east-2": Object {
        "states": "states.us-east-2.amazonaws.com",
      },
      "us-gov-east-1": Object {
        "states": "states.us-gov-east-1.amazonaws.com",
      },
      "us-gov-west-1": Object {
        "states": "states.us-gov-west-1.amazonaws.com",
      },
      "us-iso-east-1": Object {
        "states": "states.amazonaws.com",
      },
      "us-iso-west-1": Object {
        "states": "states.amazonaws.com",
      },
      "us-isob-east-1": Object {
        "states": "states.amazonaws.com",
      },
      "us-west-1": Object {
        "states": "states.us-west-1.amazonaws.com",
      },
      "us-west-2": Object {
        "states": "states.us-west-2.amazonaws.com",
      },
    },
  },
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "Test7BFAF513": Object {
      "DependsOn": Array [
        "TestRole17AB2208",
      ],
      "Properties": Object {
        "DefinitionString": "{\\"States\\":{\\"one\\":{\\"Parameters\\":{\\"thingOne\\":\\"test passed\\"}}}}",
        "RoleArn": Object {
          "Fn::GetAtt": Array [
            "TestRole17AB2208",
            "Arn",
          ],
        },
        "StateMachineName": "A-nice-state-machine",
        "StateMachineType": "STANDARD",
      },
      "Type": "AWS::StepFunctions::StateMachine",
    },
    "TestRole17AB2208": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Object {
                  "Fn::FindInMap": Array [
                    "ServiceprincipalMap",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    "states",
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
